temporal-sdk/src/main/java/io/temporal/internal/worker/LocalActivityWorker.java

```diff
@@ -180,149 +243,375 @@ public void handle(LocalActivityTask task) throws Exception {
+    public Throwable wrapFailure(LocalActivityAttemptTask task, Throwable failure) {
       return new RuntimeException("Failure processing local activity task.", failure);
     }
+  }
 
-    private @Nonnull ActivityTaskHandler.Result handleLocalActivity(
-        PollActivityTaskQueueResponse.Builder activityTask,
-        long activityStartTimeMs,
-        long localRetryThresholdMs,
-        Scope metricsScope) {
-      int attempt = activityTask.getAttempt();
+  private class LocalActivityRetryHandler implements Runnable {
+    private final LocalActivityAttemptTask localActivityAttemptTask;
 
-      metricsScope.counter(MetricsType.LOCAL_ACTIVITY_TOTAL_COUNTER).inc(1);
+    private LocalActivityRetryHandler(LocalActivityAttemptTask localActivityAttemptTask) {
+      this.localActivityAttemptTask = localActivityAttemptTask;
+    }
 
-      ActivityTaskHandler.Result result;
-      Stopwatch sw = metricsScope.timer(MetricsType.LOCAL_ACTIVITY_EXECUTION_LATENCY).start();
-      try {
-        result =
-            handler.handle(new ActivityTask(activityTask.build(), () -> {}), metricsScope, true);
-      } finally {
-        sw.stop();
-      }
-      result.setAttempt(attempt);
+    @Override
+    public void run() {
+      submitAnAttempt(localActivityAttemptTask);
+    }
+  }
 
-      if (isNonRetryableApplicationFailure(result)) {
-        return result;
-      }
+  private static class ScheduleToCloseTimeoutHandler implements Runnable {
+    private final LocalActivityExecutionContext executionContext;
 
-      if (result.getTaskCompleted() != null) {
-        com.uber.m3.util.Duration e2eDuration =
-            ProtobufTimeUtils.toM3DurationSinceNow(activityTask.getScheduledTime());
-        metricsScope.timer(MetricsType.LOCAL_ACTIVITY_SUCCEED_E2E_LATENCY).record(e2eDuration);
-      }
+    private ScheduleToCloseTimeoutHandler(LocalActivityExecutionContext executionContext) {
+      this.executionContext = executionContext;
+    }
 
-      if (result.getTaskCompleted() != null
-          || result.getTaskCanceled() != null
-          || !activityTask.hasRetryPolicy()) {
-        return result;
+    @Override
+    public void run() {
+      executionContext.callback(
+          failed(
+              executionContext.getActivityId(),
+              RetryState.RETRY_STATE_TIMEOUT,
+              newTimeoutFailure(
+                  TimeoutType.TIMEOUT_TYPE_SCHEDULE_TO_CLOSE, executionContext.getLastFailure()),
+              executionContext.getCurrentAttempt(),
+              null));
+    }
+  }
+
+  private class StartToCloseTimeoutHandler implements Runnable {
+    private final LocalActivityAttemptTask attemptTask;
+
+    private StartToCloseTimeoutHandler(LocalActivityAttemptTask attemptTask) {
+      this.attemptTask = attemptTask;
+    }
+
+    @Override
+    public void run() {
+      LocalActivityExecutionContext executionContext = attemptTask.getExecutionContext();
```

Comment:
FYI: I realized yesterday Core isn't retrying start-to-close timeouts right now, we opened a bug to do it